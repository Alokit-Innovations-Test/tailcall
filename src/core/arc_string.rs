use std::sync::Arc;

use headers::HeaderValue;
use schemars::gen::SchemaGenerator;
use schemars::schema::Schema;
use serde::{Deserialize, Deserializer, Serialize, Serializer};

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct ArcString(Arc<str>);

impl<'de> Deserialize<'de> for ArcString {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        String::deserialize(deserializer).map(ArcString::from)
    }
}

impl Serialize for ArcString {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self.as_str().serialize(serializer)
    }
}

impl Default for ArcString {
    fn default() -> Self {
        ArcString("".into())
    }
}

impl AsRef<str> for ArcString {
    fn as_ref(&self) -> &str {
        &self.0
    }
}

impl From<&str> for ArcString {
    fn from(s: &str) -> Self {
        ArcString(s.into())
    }
}

impl From<&String> for ArcString {
    fn from(s: &String) -> Self {
        ArcString(s.clone().into())
    }
}

impl From<String> for ArcString {
    fn from(s: String) -> Self {
        ArcString(s.into())
    }
}

impl From<ArcString> for Vec<u8> {
    fn from(s: ArcString) -> Self {
        s.as_str().as_bytes().to_vec()
    }
}

impl TryFrom<ArcString> for HeaderValue {
    type Error = <HeaderValue as TryFrom<String>>::Error;

    fn try_from(s: ArcString) -> Result<Self, Self::Error> {
        s.as_str().parse()
    }
}

impl PartialEq<&str> for ArcString {
    fn eq(&self, other: &&str) -> bool {
        self.as_ref() == *other
    }
}

impl schemars::JsonSchema for ArcString {
    fn schema_name() -> String {
        <String as schemars::JsonSchema>::schema_name()
    }

    fn json_schema(gen: &mut SchemaGenerator) -> Schema {
        <String as schemars::JsonSchema>::json_schema(gen)
    }
}

impl std::fmt::Display for ArcString {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        self.as_str().fmt(f)
    }
}

impl ArcString {
    pub fn as_str(&self) -> &str {
        self.as_ref()
    }
}
