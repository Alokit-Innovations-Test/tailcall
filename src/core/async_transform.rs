use std::future::Future;

use super::valid::{Valid, Validator};

/// A configuration transformer that allows us to perform various
/// transformations on the configuration before it's further processed for
/// blueprint creation.
pub trait AsyncTransform {
    type Value: Send;
    type Error;

    fn transform(
        &self,
        value: Self::Value,
    ) -> impl Future<Output = Valid<Self::Value, Self::Error>> + Send;
}

/// A suite of common operators that are available for all transformers.
pub trait AsyncTransformerOps: AsyncTransform + Send {
    fn pipe<Other: AsyncTransform<Value = Self::Value, Error = Self::Error>>(
        self,
        other: Other,
    ) -> Pipe<Self, Other>
    where
        Self: Sized,
    {
        Pipe(self, other)
    }

    fn generate(&self) -> impl Future<Output = Valid<Self::Value, Self::Error>> + Send
    where
        Self: Send + Sync,
        Self::Value: std::default::Default,
    {
        async move { self.transform(Self::Value::default()).await }
    }

    fn when(self, cond: bool) -> When<Self>
    where
        Self: Sized,
    {
        When(self, cond)
    }
}

impl<T: AsyncTransform + Sync + Send> AsyncTransformerOps for T {}

pub struct When<A>(A, bool);
impl<A: AsyncTransform + Sync + Send> AsyncTransform for When<A> {
    type Value = A::Value;
    type Error = A::Error;

    async fn transform(
        &self,
        value: Self::Value,
    ) -> Valid<Self::Value, Self::Error> {
        if self.1 {
            self.0.transform(value).await
        } else {
            Valid::succeed(value)
        }
    }
}

/// Represents a composition of two transformers.
pub struct Pipe<A, B>(A, B);

impl<A, E, X, Y> AsyncTransform for Pipe<X, Y>
where
    X: AsyncTransform<Value = A, Error = E> + Send + Sync,
    Y: AsyncTransform<Value = A, Error = E> + Send + Sync,
    A: Sync + Send,
    E: Sync + Send,
{
    type Value = A;
    type Error = E;

    async fn transform(
        &self,
        value: Self::Value,
    ) -> Valid<Self::Value, Self::Error> {
        let result = self.0.transform(value).await;
        match result.to_result() {
            Ok(result) => self.1.transform(result).await,
            Err(err) => Valid::from_validation_err(err),
        }
    }
}

/// Represents an empty transformer.
pub struct Default<A, E>(std::marker::PhantomData<(A, E)>);

impl<A: Send + Sync, E: Send + Sync> AsyncTransform for Default<A, E> {
    type Value = A;
    type Error = E;
    async fn transform(
        &self,
        value: Self::Value,
    ) -> Valid<Self::Value, Self::Error> { Valid::succeed(value) }
}

// pub fn default<A, E>() -> Default<A, E> {
//     Default(std::marker::PhantomData)
// }
