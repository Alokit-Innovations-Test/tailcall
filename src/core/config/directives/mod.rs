use lazy_static::lazy_static;

mod add_field;
mod alias;
mod cache;
mod call;
mod expr;
mod federation;
mod graphql;
mod grpc;
mod http;
mod js;
mod link;
mod modify;
mod omit;
mod protected;
mod server;
mod telemetry;
mod upstream;

pub use add_field::*;
pub use alias::*;
pub use cache::*;
pub use call::*;
pub use expr::*;
pub use federation::*;
pub use graphql::*;
pub use grpc::*;
pub use http::*;
pub use js::*;
pub use link::*;
pub use modify::*;
pub use omit::*;
pub use protected::*;
pub use server::*;
pub use telemetry::*;
pub use upstream::*;

use crate::core::directive::DirectiveCodec;

lazy_static! {
    pub static ref KNOWN_DIRECTIVES: Vec<String> = vec![
        add_field::AddField::directive_name(),
        alias::Alias::directive_name(),
        cache::Cache::directive_name(),
        call::Call::directive_name(),
        expr::Expr::directive_name(),
        graphql::GraphQL::directive_name(),
        grpc::Grpc::directive_name(),
        http::Http::directive_name(),
        js::JS::directive_name(),
        link::Link::directive_name(),
        modify::Modify::directive_name(),
        omit::Omit::directive_name(),
        protected::Protected::directive_name(),
        server::Server::directive_name(),
        telemetry::Telemetry::directive_name(),
        upstream::Upstream::directive_name(),
    ];
}
