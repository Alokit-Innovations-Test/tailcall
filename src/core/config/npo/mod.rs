mod identifier;
mod queries;

use std::{collections::HashSet, rc::Rc};

pub use identifier::*;
pub use queries::*;

use super::Config;

struct NPOIdentifier<'a> {
    config: &'a Config,
}

impl<'a> NPOIdentifier<'a> {
    pub fn new(config: &'a Config) -> NPOIdentifier {
        NPOIdentifier { config }
    }

    fn iter(
        &self,
        path: Chunk<FieldName<'a>>,
        type_name: TypeName<'a>,
        is_list: bool,
        visited: &mut HashSet<TypeName<'a>>,
    ) -> Chunk<Chunk<FieldName<'a>>> {
        if visited.contains(&type_name) {
            return Chunk::new();
        } else {
            visited.insert(type_name);
        }
        let mut chunks = Chunk::new();
        if let Some(type_of) = self.config.find_type(&type_name.as_str()) {
            for (name, field) in type_of.fields.iter() {
                let path = path.clone().append(FieldName::new(name));
                let is_batch = field.has_batched_resolver();

                if field.has_resolver() {
                    if !is_batch && is_list {
                        chunks = chunks.append(path.clone());
                    }
                }

                let is_list = is_list | field.list;

                chunks = chunks.concat(self.iter(
                    path,
                    TypeName::new(field.type_of.as_str()),
                    is_list,
                    visited,
                ))
            }
        }

        chunks
    }

    fn find_chunks(&self) -> Chunk<Chunk<FieldName<'a>>> {
        match &self.config.schema.query {
            None => Chunk::new(),
            Some(query) => self.iter(
                Chunk::new(),
                TypeName::new(query.as_str()),
                false,
                &mut HashSet::new(),
            ),
        }
    }

    fn find(&self) -> Vec<Vec<&'a str>> {
        self.find_chunks()
            .as_vec()
            .iter()
            .map(|chunk| {
                chunk
                    .as_vec()
                    .iter()
                    .map(|field_name| field_name.as_str())
                    .collect()
            })
            .collect()
    }
}

#[derive(Clone)]
enum Chunk<A> {
    Nil,
    Cons(A, Rc<Chunk<A>>),
    Concat(Rc<Chunk<A>>, Rc<Chunk<A>>),
}

impl<A> Chunk<A> {
    fn new() -> Self {
        Self::Nil
    }

    fn from_vec(vec: Vec<A>) -> Chunk<A> {
        let mut path = Chunk::new();
        for a in vec.into_iter().rev() {
            path = Chunk::Cons(a, Rc::new(path));
        }
        path
    }

    fn append(self, a: A) -> Self {
        Chunk::Cons(a, Rc::new(self))
    }

    fn concat(self, other: Chunk<A>) -> Self {
        Self::Concat(Rc::new(self), Rc::new(other))
    }

    fn as_vec(&self) -> Vec<&A> {
        let mut vec = Vec::new();
        self.as_vec_mut(&mut vec);
        vec
    }

    fn as_vec_mut<'a>(&'a self, buf: &mut Vec<&'a A>) {
        match self {
            Chunk::Nil => {}
            Chunk::Cons(a, rest) => {
                buf.push(a);
                rest.as_vec_mut(buf);
            }
            Chunk::Concat(a, b) => {
                a.as_vec_mut(buf);
                b.as_vec_mut(buf);
            }
        }
    }
}
