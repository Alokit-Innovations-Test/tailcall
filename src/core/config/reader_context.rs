use std::borrow::Cow;
use std::collections::BTreeMap;

use headers::HeaderMap;
use serde::Deserialize;
use crate::core::has_headers::HasHeaders;
use crate::core::json::JsonLike;
use crate::core::path::PathString;
use crate::core::runtime::TargetRuntime;

pub struct ConfigReaderContext<'a, Value> {
    pub runtime: &'a TargetRuntime<Value>,
    pub vars: &'a BTreeMap<String, String>,
    pub headers: HeaderMap,
}

impl<'a, Value: JsonLike<'a> + Deserialize<'a> + Clone> PathString for ConfigReaderContext<'a, Value> {
    fn path_string<T: AsRef<str>>(&self, path: &[T]) -> Option<Cow<'_, str>> {
        if path.is_empty() {
            return None;
        }

        path.split_first()
            .and_then(|(head, tail)| match head.as_ref() {
                "vars" => self.vars.get(tail[0].as_ref()).map(|v| v.into()),
                "env" => self.runtime.env.get(tail[0].as_ref()),
                _ => None,
            })
    }
}

impl<Value> HasHeaders for ConfigReaderContext<'_, Value> {
    fn headers(&self) -> &HeaderMap {
        &self.headers
    }
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;

    use super::*;
    use crate::core::tests::TestEnvIO;

    #[test]
    fn path_string() {
        let mut runtime = crate::core::runtime::test::init(None);
        runtime.env = Arc::new(TestEnvIO::from_iter([(
            "ENV_1".to_owned(),
            "ENV_VAL".to_owned(),
        )]));

        let reader_context = ConfigReaderContext {
            runtime: &runtime,
            vars: &BTreeMap::from_iter([("VAR_1".to_owned(), "VAR_VAL".to_owned())]),
            headers: Default::default(),
        };

        assert_eq!(
            reader_context.path_string(&["env", "ENV_1"]),
            Some("ENV_VAL".into())
        );
        assert_eq!(reader_context.path_string(&["env", "ENV_5"]), None);
        assert_eq!(
            reader_context.path_string(&["vars", "VAR_1"]),
            Some("VAR_VAL".into())
        );
        assert_eq!(reader_context.path_string(&["vars", "VAR_6"]), None);
        assert_eq!(reader_context.path_string(&["unknown", "unknown"]), None);
    }
}
