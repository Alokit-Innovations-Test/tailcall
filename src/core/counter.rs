use std::cell::RefCell;

#[allow(unused)]
#[derive(Default)]
pub struct Counter(RefCell<usize>);
impl Counter {
    pub fn next(&self) -> usize {
        self.0.replace_with(|a| *a + 1)
    }
}
