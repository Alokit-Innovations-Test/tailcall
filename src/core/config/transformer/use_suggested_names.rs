use std::collections::{HashMap, HashSet};

use inflector::Inflector;

use crate::core::config::Config;
use crate::core::valid::Valid;
use crate::core::Transform;

// goes through operation type names and set's it's output type name;
pub struct SuggestNames(HashSet<String>);

impl SuggestNames {
    pub fn new(suggested_names: HashSet<String>) -> Self {
        Self(suggested_names)
    }

    pub fn apply(&self, mut config: Config) -> Config {
        let _ty_names = vec![
            config.schema.query.as_ref(),
            config.schema.mutation.as_ref(),
            config.schema.subscription.as_ref(),
        ]
        .into_iter()
        .flatten()
        .collect::<Vec<_>>();

        let mut finalized_candidates = HashMap::new();
        for type_name in _ty_names {
            if let Some(type_1) = config.types.get(type_name) {
                for (field_name, field_1) in type_1.fields.iter() {
                    let singularized_name = field_name.to_singular().to_pascal_case();
                    if config.types.contains_key(&singularized_name)
                        || finalized_candidates.contains_key(&field_1.type_of)
                        || config.is_scalar(&field_1.type_of)
                        || !self.0.contains(field_name)
                    {
                        continue;
                    }
                    finalized_candidates.insert(field_1.type_of.to_owned(), singularized_name);
                }
            }
        }

        for (old_type_name, new_type_name) in finalized_candidates {
            if let Some(type_) = config.types.remove(old_type_name.as_str()) {
                // Add newly generated type.
                config.types.insert(new_type_name.to_owned(), type_);

                // Replace all the instances of old name in config.
                for actual_type in config.types.values_mut() {
                    for actual_field in actual_type.fields.values_mut() {
                        if actual_field.type_of == old_type_name {
                            // Update the field's type with the new name
                            actual_field.type_of.clone_from(&new_type_name);
                        }
                    }
                }
            }
        }

        config
    }
}

impl Transform for SuggestNames {
    type Value = Config;
    type Error = String;

    fn transform(&self, value: Self::Value) -> crate::core::valid::Valid<Self::Value, Self::Error> {
        Valid::succeed(self.apply(value))
    }
}
