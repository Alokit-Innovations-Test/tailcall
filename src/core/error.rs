use std::fmt::Display;
use std::string::FromUtf8Error;

use derive_more::From;
use inquire::InquireError;

use super::config::UnsupportedConfigFormat;

#[derive(From, thiserror::Error, Debug)]
pub enum Error {
    #[error("Worker Error: {}", _0)]
    Worker(worker::Error),

    #[error("File Error: {}", _0)]
    File(file::Error),

    #[error("Inquire Error: {}", _0)]
    Inquire(InquireError),

    #[error("Serde Yaml Error: {}", _0)]
    SerdeYaml(serde_yaml::Error),

    #[error("Package name is required")]
    PackageNameNotFound,

    #[error("Protox Parse Error")]
    ProtoxParse(protox_parse::ParseError),

    #[error("Unable to extract content of google well-known proto file")]
    GoogleProtoFileContentNotExtracted,

    #[error("Utf8 Error: {}", _0)]
    Utf8(FromUtf8Error),

    #[error("Unsupported Config Format")]
    UnsupportedConfigFormat(UnsupportedConfigFormat),

    #[error("Unsupported File Format")]
    UnsupportedFileFormat,

    #[error("Serde Json Error: {}", _0)]
    SerdeJson(serde_json::Error),

    #[error("Unable to determine path")]
    PathDeterminationFailed,

    #[error("Schema mismatch Error")]
    SchemaMismatch,

    #[error("Error: {}", _0)]
    Anyhow(anyhow::Error),
}

pub mod worker {
    use std::sync::Arc;

    use derive_more::{DebugCustom, From};
    use tokio::task::JoinError;

    #[derive(From, DebugCustom, Clone)]
    pub enum Error {
        #[debug(fmt = "Failed to initialize worker")]
        InitializationFailed,

        #[debug(fmt = "Worker communication error")]
        Communication,

        #[debug(fmt = "Serde Json Error: {}", _0)]
        SerdeJson(Arc<serde_json::Error>),

        #[debug(fmt = "Request Clone Failed")]
        RequestCloneFailed,

        #[debug(fmt = "Hyper Header To Str Error: {}", _0)]
        HyperHeaderStr(Arc<hyper::header::ToStrError>),

        #[debug(fmt = "JS Runtime Stopped Error")]
        JsRuntimeStopped,

        #[debug(fmt = "CLI Error : {}", _0)]
        CLI(String),

        #[debug(fmt = "Join Error : {}", _0)]
        Join(Arc<JoinError>),

        #[debug(fmt = "Runtime not initialized")]
        RuntimeNotInitialized,

        #[debug(fmt = "{} is not a function", _0)]
        #[from(ignore)]
        InvalidFunction(String),

        #[debug(fmt = "Rquickjs Error: {}", _0)]
        #[from(ignore)]
        Rquickjs(String),

        #[debug(fmt = "Deserialize Failed: {}", _0)]
        #[from(ignore)]
        DeserializeFailed(String),

        #[debug(fmt = "globalThis not initialized: {}", _0)]
        #[from(ignore)]
        GlobalThisNotInitialised(String),

        #[debug(
            fmt = "Error: {}\nUnable to parse value from js function: {} maybe because it's not returning a string?",
            _0,
            _1
        )]
        FunctionValueParseError(String, String),

        #[debug(fmt = "Error : {}", _0)]
        Anyhow(Arc<anyhow::Error>),
    }

    impl From<serde_json::Error> for Error {
        fn from(error: serde_json::Error) -> Self {
            Error::SerdeJson(Arc::new(error))
        }
    }

    impl From<hyper::header::ToStrError> for Error {
        fn from(error: hyper::header::ToStrError) -> Self {
            Error::HyperHeaderStr(Arc::new(error))
        }
    }

    impl From<JoinError> for Error {
        fn from(error: JoinError) -> Self {
            Error::Join(Arc::new(error))
        }
    }

    impl From<anyhow::Error> for Error {
        fn from(error: anyhow::Error) -> Self {
            Error::Anyhow(Arc::new(error))
        }
    }

    pub type Result<A> = std::result::Result<A, Error>;
}

impl Display for worker::Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            worker::Error::InitializationFailed => write!(f, "Failed to initialize worker"),
            worker::Error::Communication => write!(f, "Worker communication error"),
            worker::Error::SerdeJson(error) => write!(f, "Serde Json Error: {}", error),
            worker::Error::RequestCloneFailed => write!(f, "Request Clone Failed"),
            worker::Error::HyperHeaderStr(error) => {
                write!(f, "Hyper Header To Str Error: {}", error)
            }
            worker::Error::JsRuntimeStopped => write!(f, "JS Runtime Stopped Error"),
            worker::Error::CLI(msg) => write!(f, "CLI Error: {}", msg),
            worker::Error::Join(error) => write!(f, "Join Error: {}", error),
            worker::Error::RuntimeNotInitialized => write!(f, "Runtime not initialized"),
            worker::Error::InvalidFunction(function_name) => {
                write!(f, "{} is not a function", function_name)
            }
            worker::Error::Rquickjs(error) => write!(f, "Rquickjs error: {}", error),
            worker::Error::DeserializeFailed(error) => write!(f, "Deserialize Failed: {}", error),
            worker::Error::GlobalThisNotInitialised(error) => write!(f, "globalThis not initialized: {}", error),
            worker::Error::FunctionValueParseError(error, name) => write!(f, "Error: {}\nUnable to parse value from js function: {} maybe because it's not returning a string?", error, name),
            worker::Error::Anyhow(msg) => write!(f, "Error: {}", msg),
        }
    }
}

pub mod file {
    use std::string::FromUtf8Error;

    use derive_more::{DebugCustom, From};

    #[derive(From, DebugCustom)]
    pub enum Error {
        #[debug(fmt = "No such file or directory (os error 2)")]
        NotFound,

        #[debug(fmt = "No permission to access the file")]
        NoPermission,

        #[debug(fmt = "Access denied")]
        AccessDenied,

        #[debug(fmt = "Invalid file format")]
        InvalidFormat,

        #[debug(fmt = "Invalid file path")]
        InvalidFilePath,

        #[debug(fmt = "Invalid OS string")]
        InvalidOsString,

        #[debug(fmt = "Failed to read file : {}", _0)]
        FileReadFailed(String),

        #[debug(fmt = "Failed to write file : {}", _0)]
        #[from(ignore)]
        FileWriteFailed(String),

        #[debug(fmt = "Std IO Error: {}", _0)]
        StdIO(std::io::Error),

        #[debug(fmt = "Utf8 Error: {}", _0)]
        Utf8(FromUtf8Error),

        #[debug(fmt = "File writing not supported on Lambda.")]
        LambdaFileWriteNotSupported,

        #[debug(fmt = "Cannot write to a file in an execution spec")]
        ExecutionSpecFileWriteFailed,

        #[debug(fmt = "Cloudflare Worker Execution Error : {}", _0)]
        #[from(ignore)]
        Cloudflare(String),

        #[debug(fmt = "Error : {}", _0)]
        Anyhow(anyhow::Error),
    }

    pub type Result<A> = std::result::Result<A, Error>;
}

impl Display for file::Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            file::Error::NotFound => write!(f, "No such file or directory (os error 2)"),
            file::Error::NoPermission => write!(f, "No permission to access the file"),
            file::Error::AccessDenied => write!(f, "Access denied"),
            file::Error::InvalidFormat => write!(f, "Invalid file format"),
            file::Error::InvalidFilePath => write!(f, "Invalid file path"),
            file::Error::InvalidOsString => write!(f, "Invalid OS string"),
            file::Error::FileReadFailed(path) => write!(f, "Failed to read file: {}", path),
            file::Error::FileWriteFailed(path) => write!(f, "Failed to write file: {}", path),
            file::Error::StdIO(error) => write!(f, "Std IO Error: {}", error),
            file::Error::Utf8(error) => write!(f, "Utf8 Error: {}", error),
            file::Error::LambdaFileWriteNotSupported => {
                write!(f, "File writing not supported on Lambda.")
            }
            file::Error::ExecutionSpecFileWriteFailed => {
                write!(f, "Cannot write to a file in an execution spec")
            }
            file::Error::Cloudflare(error) => {
                write!(f, "Cloudflare Worker Execution Error: {}", error)
            }
            file::Error::Anyhow(msg) => write!(f, "Error: {}", msg),
        }
    }
}

pub type Result<A, E> = std::result::Result<A, E>;
