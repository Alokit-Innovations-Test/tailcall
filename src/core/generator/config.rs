use std::path::Path;

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use crate::core::config::{self, reader::ConfigReader};

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub enum InputSource {
    /// src: maintains the same src written in config.
    /// resolved_src: holds the correctly resolved src with respect to config.
    Config {
        src: String,
        resolved_src: Option<String>,
    },
    Import {
        src: String,
        resolved_src: Option<String>,
    },
}

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Input {
    #[serde(flatten)]
    pub source: InputSource,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Output {
    /// Controls the output format (graphql, json, yaml)
    pub format: config::Source,
    /// Specifies the output file name
    pub file: String,
}

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Schema {
    #[serde(default = "defaults::schema::query")]
    pub query: String,
    #[serde(default = "defaults::schema::mutation")]
    pub mutation: String,
}

impl Default for Schema {
    fn default() -> Self {
        Self {
            query: defaults::schema::query(),
            mutation: defaults::schema::mutation(),
        }
    }
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct GenerateOptions {
    #[serde(default)]
    pub schema: Schema,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct Transform {
    // TODO: change types
    pub ambiguous_name_resolver: Option<serde_json::Value>,
    pub merge_types: Option<serde_json::Value>,
    pub js: Option<serde_json::Value>,
}

#[derive(Serialize, Deserialize, Default, Debug, JsonSchema)]
#[serde(rename_all = "camelCase")]
pub struct GeneratorConfig {
    pub input: Vec<Input>,
    #[serde(default)]
    pub output: Output,
    #[serde(default)]
    pub generate: GenerateOptions,
    #[serde(default)]
    pub transform: Transform,
}

impl GeneratorConfig {
    /// Resolves all the paths present inside the GeneratorConfig.
    pub fn resolve_paths(mut self, file_path: &str) -> Self {
        let config_path = Path::new(file_path).parent().unwrap_or(Path::new(""));

        for input in self.input.iter_mut() {
            match &mut input.source {
                InputSource::Config { src, resolved_src }
                | InputSource::Import { src, resolved_src } => {
                    *resolved_src = Some(ConfigReader::resolve_path(src, Some(config_path)));
                }
            }
        }

        self.output.file = ConfigReader::resolve_path(&self.output.file, Some(config_path));

        self
    }
}

mod defaults {
    pub mod schema {
        pub fn query() -> String {
            "Query".to_string()
        }

        pub fn mutation() -> String {
            "Mutation".to_string()
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::GeneratorConfig;

    #[test]
    fn test_from_json() {
        let content = std::fs::read_to_string(tailcall_fixtures::generator::SIMPLE_JSON).unwrap();
        let config: GeneratorConfig = serde_json::from_str(&content).unwrap();

        assert_debug_snapshot!(&config);
    }

    #[test]
    fn test_resolve_paths() {
        let file_path = "../../../tailcall-fixtures/fixtures/generator/simple-json.json";
        let content = std::fs::read_to_string(tailcall_fixtures::generator::SIMPLE_JSON).unwrap();
        let config: GeneratorConfig = serde_json::from_str(&content).unwrap();
        let config = config.resolve_paths(file_path);
        assert_debug_snapshot!(&config);
    }
}
