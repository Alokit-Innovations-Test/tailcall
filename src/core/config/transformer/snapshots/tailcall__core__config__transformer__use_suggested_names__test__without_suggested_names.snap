---
source: src/core/config/transformer/use_suggested_names.rs
expression: transformed_config.to_sdl()
---
schema @server(port: 8000) @upstream(baseURL: "http://jsonplaceholder.typicode.com", batch: {delay: 100, headers: []}, httpCache: 42) {
  query: Query
}

type Query {
  post(id: ID!): T1 @http(path: "/posts")
  todos: [T3] @http(path: "/todos")
  users: [T2] @http(path: "/users")
}

type T1 {
  body: String
  id: ID
  title: String
}

type T2 {
  id: ID
  name: String
  relatedPosts: [T1] @http(path: "/posts", query: [{key: "id", value: "{{.value.id}}"}])
}

type T3 {
  id: ID
  todoPost: T1
}
